clang/lib/Sema/SemaExpr.cpp

```diff
@@ -9124,6 +9152,43 @@ QualType Sema::GetCheckedCInteropType(ExprResult LHS) {
     return QualType();
 }
 
+/// Get the bounds-safe interface type for an rvalue expression, if the
+/// rvalue expression has a bounds-safe interface. Return a null QualType
+/// otherwise.  The rvalue expression may appear as part of the left-hand
+/// side of an assignment - for example, as the subexpression of a pointer
+/// deference or an array subscript.  For rvalue expressions appearing as
+/// part of the left-hand side of an assignment, only lvalue-to-rvalue casts
+/// and pointer arithmetic have bounds-safe interfaces.
+QualType Sema::GetCheckedCRValueInteropType(ExprResult RHS) {
```

Comment:
In addition to returning the bounds-safe interface type for `*e` and `p +/- i`, should this method also return the bounds-safe interface type for `&e`? If `e` has bounds-safe interface type `T`, should `&e` have bounds-safe interface type `ptr<T>`?

As an example, making this change would allow the following to compile:
```
void address_of(int *interop_ptr : itype(ptr<int>), ptr<int> checked_ptr) {
  *(&interop_ptr) = checked_ptr;
}
```
Currently, this example does not compile since `&interop_ptr` has an unchecked pointer type `int **`. If GetCheckedCRValueInteropType were to return the bounds-safe interface type for address-of expressions, then `&interop_ptr` would have a bounds-safe interface type of `ptr<ptr<int>>` and so `*(&interop_ptr)` would have a bounds-safe interface type of `ptr<int>`.